### PR TITLE
[SPARK-36418][SQL] Use CAST in parsing of dates/timestamps with default pattern

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to use the `CAST` logic when the pattern is not specified in `DateFormatter` or `TimestampFormatter`. In particular, invoke the `DateTimeUtils.stringToTimestampAnsi()` or `stringToDateAnsi()` in the case.

### Why are the changes needed?
1. This can improve user experience with Spark SQL by making the default date/timestamp parsers more flexible and tolerant to their inputs.
2. We make the default case consistent to the behavior of the `CAST` expression which makes implementation more consistent.

### Does this PR introduce _any_ user-facing change?
The changes shouldn't introduce behavior change in regular cases but it can influence on corner cases. New implementation is able to parse more dates/timestamps by default. For instance, old (current) date parses can recognize dates only in the format **yyyy-MM-dd** but new one can handle:
   * `[+-]yyyy*`
   * `[+-]yyyy*-[m]m`
   * `[+-]yyyy*-[m]m-[d]d`
   * `[+-]yyyy*-[m]m-[d]d `
   * `[+-]yyyy*-[m]m-[d]d *`
   * `[+-]yyyy*-[m]m-[d]dT*`

Similarly for timestamps. The old (current) timestamp formatter is able to parse timestamps only in the format **yyyy-MM-dd HH:mm:ss** by default, but new implementation can handle:
   * `[+-]yyyy*`
   * `[+-]yyyy*-[m]m`
   * `[+-]yyyy*-[m]m-[d]d`
   * `[+-]yyyy*-[m]m-[d]d `
   * `[+-]yyyy*-[m]m-[d]d [h]h:[m]m:[s]s.[ms][ms][ms][us][us][us][zone_id]`
   * `[+-]yyyy*-[m]m-[d]dT[h]h:[m]m:[s]s.[ms][ms][ms][us][us][us][zone_id]`
   * `[h]h:[m]m:[s]s.[ms][ms][ms][us][us][us][zone_id]`
   * `T[h]h:[m]m:[s]s.[ms][ms][ms][us][us][us][zone_id]`

### How was this patch tested?
By running the affected test suites:
```
$ build/sbt "test:testOnly *ImageFileFormatSuite"
$ build/sbt "test:testOnly *ParquetV2PartitionDiscoverySuite"
```

### CODE REVIEW

**Comment by MaxGekk** [comment]
@cloud-fan @gengliangwang Please, review this PR.

File: mllib/src/test/scala/org/apache/spark/ml/source/image/ImageFileFormatSuite.scala [hunk 1]
```diff
@@ -95,14 +96,14 @@ class ImageFileFormatSuite extends SparkFunSuite with MLlibTestSparkContext {
       .collect()
 
     assert(Set(result: _*) === Set(
-      Row("29.5.a_b_EGDP022204.jpg", "kittens", "2018-01"),
```
Comment (gengliangwang) [inline mllib/src/test/scala/org/apache/spark/ml/source/image/ImageFileFormatSuite.scala:98]: why do we need the changes here?

Comment (MaxGekk) [reply mllib/src/test/scala/org/apache/spark/ml/source/image/ImageFileFormatSuite.scala:98]: After the changes, the partition value `2018-01` became a valid date value. New partition formatter can parse it as 2018-01-01, see patterns that supported by the CAST expression:
https://github.com/apache/spark/blob/ed0e351f05ac6edc132c3a630206b2031c419e1c/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala#L228
As a consequence, Spark infers DateType as the type of partition values. And finally, it converts all strings to the type.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetPartitionDiscoverySuite.scala [hunk 2]
```diff
@@ -1067,7 +1067,7 @@ abstract class ParquetPartitionDiscoverySuite
 
   test("SPARK-23436: invalid Dates should be inferred as String in partition inference") {
     withTempPath { path =>
-      val data = Seq(("1", "2018-01", "2018-01-01-04", "test"))
+      val data = Seq(("1", "2018-41", "2018-01-01-04", "test"))
```
Comment (wayne-kyungwonpark) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetPartitionDiscoverySuite.scala:1070]: @MaxGekk 
Hello, 
It's late, but I'd like to ask the 'my' test result related to this test code.

In spark 3.3.0, I tested what is the read type of partition columns.
When the date_month column value was '2018-01-01', I was able to get the `DateType` type, but I couldn't obtain it when the value was '2018-01' like below.
```scala
val data = Seq(("1", "2018-01", "2018-01-01-04", "test"))
  .toDF("id", "date_month", "date_hour", "data")
val path = "some_path"
data.write.partitionBy("date_month", "date_hour").parquet(path)
val input = spark.read.parquet(path).select("id", "date_month", "date_hour", "data")
println(input.schema)
println(data.schema)
``` 

```
StructType(StructField(id,StringType,true), StructField(date_month,StringType,true), StructField(date_hour,StringType,true), StructField(data,StringType,true))
StructType(StructField(id,StringType,true), StructField(date_month,StringType,true), StructField(date_hour,StringType,true), StructField(data,StringType,true))
```

So, I would like to ask what it means to change the date_month value from '2018-01' to '2018-41' in this test code.
Is there anything I might be misunderstanding or any mistakes in my testing?

Thanks.🙏

Comment (wayne-kyungwonpark) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetPartitionDiscoverySuite.scala:1070]: Additionally,
I tested this code using sbt by replacing the value of the date_month column as shown below.
(`2018-04` -> `2018-01`)
```
test(
  "SPARK-23436, SPARK-36861: invalid Dates should be inferred as String in partition inference") {
  withTempPath { path =>
    val data = Seq(("1", "2018-01", "2018-01-01-04", "2021-01-01T00", "test"))
      .toDF("id", "date_month", "date_hour", "date_t_hour", "data")

    data.write.partitionBy("date_month", "date_hour", "date_t_hour").parquet(path.getAbsolutePath)
    val input = spark.read.parquet(path.getAbsolutePath).select("id",
      "date_month", "date_hour", "date_t_hour", "data")

    assert(data.schema.sameType(input.schema))
    checkAnswer(input, data)
  }
}
```
```sbt
sbt:spark-sql> testOnly *ParquetV1PartitionDiscoverySuite -- -z "SPARK-23436"
sbt:spark-sql> testOnly *ParquetV2PartitionDiscoverySuite -- -z "SPARK-23436"
```
=> All success..!

However, I changed the value to `2018-01-01` with 'yyyy-MM-dd' `DateType` format,
sbt tests are all failed!

**Pull request closed without merging**